cfgov/research_reports/models.py

```diff
@@ -1,3 +1,142 @@
+    FieldPanel, InlinePanel, MultiFieldPanel, ObjectList, TabbedInterface
+)
+from wagtail.admin.forms.models import WagtailAdminModelFormMetaclass
+from wagtail.admin.forms.pages import WagtailAdminPageForm
+from wagtail.core.models import PageManager
+
+from modelcluster.fields import ParentalKey
+from modelcluster.models import ClusterableModel
+
+from v1.models import SublandingFilterablePage
+from v1.models.base import CFGOVPage
+
+
+def get_toc_sections(request, page):
+    return [{
+        'expanded': False,
+        'title': section.header,
+        'url': '#' + section.html_id,
+        'children': [{
+            'title': subsection.sub_header, 'url': '#' + subsection.sub_id
+        } for subsection in section.report_subsections.all().order_by('pk')]
+    } for section in
+        page.report_sections.all().order_by('pk')
+        if not section.is_appendix]
+
+
+def get_toc_appendices(request, page):
+    return [{
+        'expanded': False,
+        'title': section.header,
+        'url': '#' + section.html_id,
+        'children': [{
+            'title': subsection.sub_header, 'url': '#' + subsection.sub_id
+        } for subsection in section.report_subsections.all().order_by('pk')]
+    } for section in
+        page.report_sections.all().order_by('pk')
+        if section.is_appendix]
+
+
+def get_researchers():
+    return dict([
+        (r.title, r.url) for r in
+        SublandingFilterablePage.objects.get(pk=4833).get_children()
+    ])
+
+
+class ReportSection(ClusterableModel):
+    header = models.CharField(max_length=200, blank=True)
+    html_id = models.CharField(max_length=50, blank=True)
+    is_appendix = models.BooleanField(default=False)
+    body = models.TextField(blank=True)
+    panels = [
+        FieldPanel('header'),
+        FieldPanel('html_id'),
+        FieldPanel('is_appendix'),
+        FieldPanel('body'),
+        InlinePanel('report_subsections', label='Subsection'),
+    ]
+    action = ParentalKey('Report',
+                         on_delete=models.CASCADE,
+                         related_name='report_sections')
+
+
+class ReportSubSection(models.Model):
+    sub_header = models.CharField(max_length=200)
+    sub_id = models.CharField(max_length=50, blank=True)
+    sub_body = models.TextField(blank=True)
+    action = ParentalKey('ReportSection',
+                         on_delete=models.CASCADE,
+                         related_name='report_subsections')
+
+
+class AuthorNames(models.Model):
+    name = models.CharField(max_length=50)
+    action = ParentalKey('Report',
+                         on_delete=models.CASCADE,
+                         related_name='report_author_names')
+
+
+class ReportMetaclass(WagtailAdminModelFormMetaclass):
+    @classmethod
+    def child_form(cls):
+        return ReportForm
+
+
+class ReportForm(WagtailAdminPageForm, metaclass=ReportMetaclass):
+    pass
+
+
+class Report(CFGOVPage):
+    base_form_class = ReportForm
```

Comment:
What does this do?

Reply:
It's a workaround that allows nested inline panels. See https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/models/home_page.py#L20-L24 (and the linked issues for more gory details).